h/services/job_queue.py

```diff
@@ -0,0 +1,170 @@
+
+       Jobs are processed in batch by a periodic Celery task that runs every N
+       minutes (see h-periodic for how frequently the task is run) so a job
+       added to the job queue might not get processed until N minutes later (or
+       even longer: if the job queue is currently long then the periodic task
+       may have to run multiple times before it gets to your job).
+
+       In contrast tasks added to Celery can be processed by a worker almost
+       immediately.
+
+    3. The job queue is very simple and has far fewer features than Celery.
+
+    Celery should be the default task queue for almost all tasks, and only jobs
+    that really need Postgres transactionality should use this custom job
+    queue.
+
+    At the time of writing this job queue only supports a single type of job:
+    synchronizing annotations from Postgres to Elasticsearch.
+    """
+
+    def __init__(self, db, es, batch_indexer, limit):
+        self._db = db
+        self._es = es
+        self._batch_indexer = batch_indexer
+        self._limit = limit
+
+    def add_sync_annotation_job(self, annotation_id, tag, scheduled_at=None):
+        """Add an annotation to be synced to Elasticsearch."""
+        self.add_sync_annotation_jobs([annotation_id], tag, scheduled_at)
+
+    def add_sync_annotation_jobs(self, annotation_ids, tag, scheduled_at=None):
+        """Add a list of annotations to the queue to be synced to Elasticsearch."""
+        self._db.add_all(
+            Job(
+                tag=tag,
+                scheduled_at=scheduled_at,
+                kwargs={"annotation_id": annotation_id},
+            )
+            for annotation_id in annotation_ids
+        )
+
+    def sync_annotations(self):
+        """
+        Synchronize annotations from Postgres to Elasticsearch.
+
+        This method is meant to be run periodically. It's called by a Celery
+        task that's scheduled periodically by h-periodic.
+
+        Each time it runs the method consumes a fixed number of sync annotation
+        jobs from the queue and copies those annotations from Postgres to
+        Elasticsearch (if they aren't in Elasticsearch already).
+
+        """
+        jobs = self._get_jobs_from_queue()
+
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
+        annotations_from_db = self._get_annotations_from_db(annotation_ids)
+        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+
+        deleted = []
+        to_index = []
+        succeeded = []
+
+        for job in jobs:
+            annotation_id = job.kwargs["annotation_id"]
+            annotation_from_db = annotations_from_db.get(annotation_id)
+            annotation_from_es = annotations_from_es.get(annotation_id)
+
+            if not annotation_from_db:
+                deleted.append(job)
+            elif annotation_from_db.deleted:
+                deleted.append(job)
```

Comment:
This could be simplified by changing the SQL query below to filter out annotations marked as deleted in the first place